In [33]:
import pandas as pd
import numpy as np

In [2]:
train_df = pd.read_csv('./Train_Data.csv')
test_df = pd.read_csv('./Test_Data.csv')
sample_submission = pd.read_csv('./Sample_Submission.csv')

## **1. Basic Data Exploration:**

In [6]:
train_shape = train_df.shape
test_shape = test_df.shape
submission_shape = sample_submission.shape

In [7]:
train_shape, test_shape, submission_shape

((22584, 19), (3986, 19), (3986, 1))

In [9]:
train_df.head()

pc      ld    m0    m1    m2       m3        m4       m5        m6       m7  \
0  y   83.71  63.0   7.2  52.5  14.0232  130.8300  12.4280  188.8425   8.9520   
1  y  108.94  31.5  12.8  84.0  13.2840  128.8350  13.5256  183.0990   8.8520   
2  E  169.65   0.0   5.6  73.5  14.5472  128.9295  13.6424  174.4680   8.9800   
3  x  122.42  31.5   7.2  63.0  15.0152  119.8575  12.3344  186.8580  10.7208   
4  E  125.43  94.5   7.2  42.0  14.4176  135.4290  14.5824  187.8135   9.3088   

         m8       m9       m10      m11       m12      m13       m14   ma  \
0  201.1905   9.2896  141.9075  16.0968  150.3390  12.4880  173.1240  m78   
1  207.2385   8.4704  154.7805  13.3304  101.0205  12.5096  131.4075  m78   
2  190.3125  11.3056  156.7650      NaN  122.5350  11.7136  176.8200  m76   
3  193.8195  10.6096  175.7490      NaN  124.8030  13.8424  168.2625  m55   
4  203.1540   9.8280  172.7040  14.4720  120.2145      NaN  150.1185  m76   

   pred  
0     0  
1     0  
2     1  
3     0  
4     0

In [10]:
test_df.head()

index pc      ld     m0    m1     m2       m3        m4       m5        m6  \
0      0  E  156.46   73.5   8.8  105.0  13.2768  113.5365  13.9552  168.5460   
1      1  y   58.87   84.0  12.8   42.0  14.4440  126.6930  13.6688  190.6275   
2      2  w  139.68  126.0   3.2   73.5  14.2432  107.4885  13.8544  168.3255   
3      3  z  194.06   52.5   6.4  105.0  14.5480  138.6105  13.5080  196.6545   
4      4  x  173.14   42.0   8.0  136.5  14.9496  103.8765  13.8936  168.3570   

        m7        m8       m9       m10      m11       m12      m13       m14  \
0   8.5896  187.6560   8.3472  178.5945  14.3192  138.0435  12.7008  166.1100   
1  10.1712  202.5870   8.5416  166.9815  15.5944  142.5690  11.6832  177.7230   
2  10.2504  214.2735  11.3392  155.1795  16.0256  132.1005  13.0104  185.6085   
3  10.6304  213.4650   9.9224  191.1000  14.8256  128.0475  12.1984  180.2745   
4   9.6184  191.2050   9.9400  185.0310  13.8664  124.7610      NaN  157.6470   

    ma  
0  m76  
1  m78  
2  m78  
3  m75  
4  m55

In [8]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22584 entries, 0 to 22583
Data columns (total 19 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   pc      22584 non-null  object 
 1   ld      22362 non-null  float64
 2   m0      22584 non-null  float64
 3   m1      22584 non-null  float64
 4   m2      22584 non-null  float64
 5   m3      22270 non-null  float64
 6   m4      22117 non-null  float64
 7   m5      22008 non-null  float64
 8   m6      21909 non-null  float64
 9   m7      21796 non-null  float64
 10  m8      21684 non-null  float64
 11  m9      21537 non-null  float64
 12  m10     21484 non-null  float64
 13  m11     21357 non-null  float64
 14  m12     21203 non-null  float64
 15  m13     21085 non-null  float64
 16  m14     20995 non-null  float64
 17  ma      22584 non-null  object 
 18  pred    22584 non-null  int64  
dtypes: float64(16), int64(1), object(2)
memory usage: 3.3+ MB


In [16]:
train_features = train_df.drop(['pred'], axis=1)
test_features = test_df.drop(['index'], axis=1)
train_labels = train_df['pred']

In [17]:
test_split_index = train_shape[0]
test_split_index

22584

In [18]:
features_data = pd.concat([train_features, test_features], axis = 0)

In [19]:
features_data

pc      ld     m0    m1     m2       m3        m4       m5        m6  \
0     y   83.71   63.0   7.2   52.5  14.0232  130.8300  12.4280  188.8425   
1     y  108.94   31.5  12.8   84.0  13.2840  128.8350  13.5256  183.0990   
2     E  169.65    0.0   5.6   73.5  14.5472  128.9295  13.6424  174.4680   
3     x  122.42   31.5   7.2   63.0  15.0152  119.8575  12.3344  186.8580   
4     E  125.43   94.5   7.2   42.0  14.4176  135.4290  14.5824  187.8135   
...  ..     ...    ...   ...    ...      ...       ...      ...       ...   
3981  z  109.45  126.0   4.0   94.5  14.6168  131.4495  13.1032  189.0840   
3982  w  188.31  136.5   1.6   21.0  12.9032  115.1535  14.3616  189.4830   
3983  E   80.99   31.5   4.8  115.5  15.0696  114.6810  14.0200  186.0810   
3984  w  120.64   42.0   0.8  126.0  13.8960  123.6585  13.5496  187.8975   
3985  w  115.73  199.5   4.0   10.5  14.2592  129.0555  13.8904  179.6865   

           m7        m8       m9       m10      m11       m12      m13  \
0      8.9520  201.1905   9.2896  141.9075  16.0968  150.3390  12.4880   
1      8.8520  207.2385   8.4704  154.7805  13.3304  101.0205  12.5096   
2      8.9800  190.3125  11.3056  156.7650      NaN  122.5350  11.7136   
3     10.7208  193.8195  10.6096  175.7490      NaN  124.8030  13.8424   
4      9.3088  203.1540   9.8280  172.7040  14.4720  120.2145      NaN   
...       ...       ...      ...       ...      ...       ...      ...   
3981   8.3384  203.2275   9.6848       NaN  14.6384   97.2510  12.3336   
3982   7.6680  183.8865   9.8032  172.2945  16.1512  110.7330  12.1680   
3983   9.2080  193.6200   9.1600  170.1840      NaN  127.4700  13.5016   
3984   9.7944  189.0525   9.4752  173.0295  17.1728  113.0430  13.1480   
3985   8.8712  214.6410   8.8816  167.6535  15.4640  136.3635  11.8728   

           m14   ma  
0     173.1240  m78  
1     131.4075  m78  
2     176.8200  m76  
3     168.2625  m55  
4     150.1185  m76  
...        ...  ...  
3981  180.0435  m75  
3982  204.5505  m78  
3983       NaN  m76  
3984  164.2410  m78  
3985  178.5420  m78  

[26570 rows x 18 columns]

In [20]:
features_data.reset_index(drop=True)

pc      ld     m0    m1     m2       m3        m4       m5        m6  \
0      y   83.71   63.0   7.2   52.5  14.0232  130.8300  12.4280  188.8425   
1      y  108.94   31.5  12.8   84.0  13.2840  128.8350  13.5256  183.0990   
2      E  169.65    0.0   5.6   73.5  14.5472  128.9295  13.6424  174.4680   
3      x  122.42   31.5   7.2   63.0  15.0152  119.8575  12.3344  186.8580   
4      E  125.43   94.5   7.2   42.0  14.4176  135.4290  14.5824  187.8135   
...   ..     ...    ...   ...    ...      ...       ...      ...       ...   
26565  z  109.45  126.0   4.0   94.5  14.6168  131.4495  13.1032  189.0840   
26566  w  188.31  136.5   1.6   21.0  12.9032  115.1535  14.3616  189.4830   
26567  E   80.99   31.5   4.8  115.5  15.0696  114.6810  14.0200  186.0810   
26568  w  120.64   42.0   0.8  126.0  13.8960  123.6585  13.5496  187.8975   
26569  w  115.73  199.5   4.0   10.5  14.2592  129.0555  13.8904  179.6865   

            m7        m8       m9       m10      m11       m12      m13  \
0       8.9520  201.1905   9.2896  141.9075  16.0968  150.3390  12.4880   
1       8.8520  207.2385   8.4704  154.7805  13.3304  101.0205  12.5096   
2       8.9800  190.3125  11.3056  156.7650      NaN  122.5350  11.7136   
3      10.7208  193.8195  10.6096  175.7490      NaN  124.8030  13.8424   
4       9.3088  203.1540   9.8280  172.7040  14.4720  120.2145      NaN   
...        ...       ...      ...       ...      ...       ...      ...   
26565   8.3384  203.2275   9.6848       NaN  14.6384   97.2510  12.3336   
26566   7.6680  183.8865   9.8032  172.2945  16.1512  110.7330  12.1680   
26567   9.2080  193.6200   9.1600  170.1840      NaN  127.4700  13.5016   
26568   9.7944  189.0525   9.4752  173.0295  17.1728  113.0430  13.1480   
26569   8.8712  214.6410   8.8816  167.6535  15.4640  136.3635  11.8728   

            m14   ma  
0      173.1240  m78  
1      131.4075  m78  
2      176.8200  m76  
3      168.2625  m55  
4      150.1185  m76  
...         ...  ...  
26565  180.0435  m75  
26566  204.5505  m78  
26567       NaN  m76  
26568  164.2410  m78  
26569  178.5420  m78  

[26570 rows x 18 columns]

In [21]:
features_data[test_split_index:]

pc      ld     m0    m1     m2       m3        m4       m5        m6  \
0     E  156.46   73.5   8.8  105.0  13.2768  113.5365  13.9552  168.5460   
1     y   58.87   84.0  12.8   42.0  14.4440  126.6930  13.6688  190.6275   
2     w  139.68  126.0   3.2   73.5  14.2432  107.4885  13.8544  168.3255   
3     z  194.06   52.5   6.4  105.0  14.5480  138.6105  13.5080  196.6545   
4     x  173.14   42.0   8.0  136.5  14.9496  103.8765  13.8936  168.3570   
...  ..     ...    ...   ...    ...      ...       ...      ...       ...   
3981  z  109.45  126.0   4.0   94.5  14.6168  131.4495  13.1032  189.0840   
3982  w  188.31  136.5   1.6   21.0  12.9032  115.1535  14.3616  189.4830   
3983  E   80.99   31.5   4.8  115.5  15.0696  114.6810  14.0200  186.0810   
3984  w  120.64   42.0   0.8  126.0  13.8960  123.6585  13.5496  187.8975   
3985  w  115.73  199.5   4.0   10.5  14.2592  129.0555  13.8904  179.6865   

           m7        m8       m9       m10      m11       m12      m13  \
0      8.5896  187.6560   8.3472  178.5945  14.3192  138.0435  12.7008   
1     10.1712  202.5870   8.5416  166.9815  15.5944  142.5690  11.6832   
2     10.2504  214.2735  11.3392  155.1795  16.0256  132.1005  13.0104   
3     10.6304  213.4650   9.9224  191.1000  14.8256  128.0475  12.1984   
4      9.6184  191.2050   9.9400  185.0310  13.8664  124.7610      NaN   
...       ...       ...      ...       ...      ...       ...      ...   
3981   8.3384  203.2275   9.6848       NaN  14.6384   97.2510  12.3336   
3982   7.6680  183.8865   9.8032  172.2945  16.1512  110.7330  12.1680   
3983   9.2080  193.6200   9.1600  170.1840      NaN  127.4700  13.5016   
3984   9.7944  189.0525   9.4752  173.0295  17.1728  113.0430  13.1480   
3985   8.8712  214.6410   8.8816  167.6535  15.4640  136.3635  11.8728   

           m14   ma  
0     166.1100  m76  
1     177.7230  m78  
2     185.6085  m78  
3     180.2745  m75  
4     157.6470  m55  
...        ...  ...  
3981  180.0435  m75  
3982  204.5505  m78  
3983       NaN  m76  
3984  164.2410  m78  
3985  178.5420  m78  

[3986 rows x 18 columns]

## **2. Data Preprocessing:**

### **2.1 | Handling Missing Values:**

In [25]:
features_data.isnull().sum()

pc        0
ld      250
m0        0
m1        0
m2        0
m3      381
m4      538
m5      676
m6      796
m7      937
m8     1048
m9     1227
m10    1300
m11    1468
m12    1601
m13    1774
m14    1874
ma        0
dtype: int64

In [29]:
features_data_filled = features_data.fillna(features_data.mean(numeric_only=True))

In [32]:
features_data_filled.isnull().sum().sum()

0

As we can see we have removed all the missing values by imputing them with mean. 

In [42]:
df = features_data_filled.copy()

### **2.2 | Handling Categorical Variables:**

### **2.3 | Data Normalisation and Feature Scaling:**

In [43]:
from sklearn.preprocessing import StandardScaler, Normalizer

scaler_standard = StandardScaler()
df_standard = scaler_standard.fit_transform(df)
df_standard = pd.DataFrame(df_standard, columns=df.columns)

# Scaling to unit norm (L2 normalization)
scaler_norm = Normalizer(norm='l2')
df_norm = scaler_norm.fit_transform(df)
df_norm = pd.DataFrame(df_norm, columns=df.columns)

ValueError: could not convert string to float: 'y'

## **3.Feature Engineering:** 
Extract relevant features from the raw data that can improve the performance of the classification algorithm. This can include techniques such as dimensionality reduction (e.g., PCA) or creating new features based on domain knowledge.

### **3.1 | Feature Exploration:**

### **3.2 | Feature Selection:**

### **3.3 | Dimensionality Reduction :**

### **3.4 | Feature Importance Evaluation:**

## **4. Splitting Dataset:**

### **4.1 | Defining the split:**

### **4.2 | Randomize Training and Validation Set:**